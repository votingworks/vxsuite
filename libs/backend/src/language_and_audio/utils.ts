import { sha256 } from 'js-sha256';
import { assert, assertDefined } from '@votingworks/basics';
import {
  UiStringAudioIdsPackage,
  UiStringsPackage,
  LanguageCode,
  isLanguageCode,
} from '@votingworks/types';

/**
 * i18next catalog strings can contain tags that interfere with speech synthesis, e.g.
 * "Do you prefer <1>apple pie</1> or <3>orange marmalade</3>?". This function cleans text in
 * preparation for speech synthesis accordingly.
 */
export function cleanText(text: string): string {
  return text.replace(/<\/?\d+>/g, '');
}

/**
 * Segment of text that is either interpolated or not
 */
export interface Segment {
  content: string;
  isInterpolated: boolean;
}

/**
 * Splits interpolated text in preparation for speech synthesis. See tests for sample inputs and
 * outputs.
 */
export function splitInterpolatedText(text: string): Segment[] {
  const interpolationRegex = /\{\{.*?\}\}/g;
  const nonInterpolatedSegments = text.split(interpolationRegex);
  const interpolatedSegments = text.match(interpolationRegex);
  const segments: Segment[] = [];
  for (const [i, nonInterpolatedSegment] of nonInterpolatedSegments.entries()) {
    segments.push({ content: nonInterpolatedSegment, isInterpolated: false });
    if (interpolatedSegments && i < interpolatedSegments.length) {
      segments.push({
        content: assertDefined(interpolatedSegments[i]),
        isInterpolated: true,
      });
    }
  }

  const segmentsCleaned = segments.map(({ content, isInterpolated }) => ({
    content: content.trim(),
    isInterpolated,
  }));

  // Allow non-empty, single-segment strings (like punctuation) through without
  // further cleaning:
  if (
    segmentsCleaned.length === 1 &&
    segmentsCleaned[0] &&
    segmentsCleaned[0].content.length > 0
  ) {
    return segmentsCleaned;
  }

  // Remove extraneous non-word segments from multi-segment strings:
  return segmentsCleaned.filter(({ content }) => /[a-z0-9]/i.test(content));
}

/**
 * Prepares text for speech synthesis by cleaning it, splitting it if interpolated, and generating
 * audio IDs for the resulting segments
 */
export function prepareTextForSpeechSynthesis(
  languageCode: LanguageCode,
  text: string
): Array<{ audioId: string; segment: Segment }> {
  return splitInterpolatedText(cleanText(text)).map((segment) => ({
    audioId: segment.isInterpolated
      ? segment.content
      : sha256([languageCode, segment.content].join(':')).slice(0, 10),
    segment,
  }));
}

/**
 * Sets a UI string in a UI strings package
 */
export function setUiString(
  uiStrings: UiStringsPackage,
  languageCode: LanguageCode,
  stringKey: string | [string, string],
  stringInLanguage: string
): void {
  uiStrings[languageCode] ??= {}; // eslint-disable-line no-param-reassign
  const uiStringsInLanguage = assertDefined(uiStrings[languageCode]);

  // Single-value key
  if (typeof stringKey === 'string') {
    uiStringsInLanguage[stringKey] = stringInLanguage;
    return;
  }

  // Two-value key
  uiStringsInLanguage[stringKey[0]] ??= {};
  const subStructure = uiStringsInLanguage[stringKey[0]];
  assert(subStructure !== undefined && typeof subStructure === 'object');
  subStructure[stringKey[1]] = stringInLanguage;
}

/**
 * Sets audio IDs for a UI string in a UI string audio IDs package
 */
export function setUiStringAudioIds(
  uiStringAudioIds: UiStringAudioIdsPackage,
  languageCode: LanguageCode,
  stringKey: string | [string, string],
  audioIds: string[]
): void {
  uiStringAudioIds[languageCode] ??= {}; // eslint-disable-line no-param-reassign
  const uiStringAudioIdsInLanguage = assertDefined(
    uiStringAudioIds[languageCode]
  );

  // Single-value key
  if (typeof stringKey === 'string') {
    uiStringAudioIdsInLanguage[stringKey] = audioIds;
    return;
  }

  // Two-value key
  uiStringAudioIdsInLanguage[stringKey[0]] ??= {};
  const subStructure = uiStringAudioIdsInLanguage[stringKey[0]];
  assert(subStructure !== undefined && !Array.isArray(subStructure));
  subStructure[stringKey[1]] = audioIds;
}

/**
 * Helper function to iterate over all UI strings in a UI strings package
 */
export function forEachUiString(
  uiStrings: UiStringsPackage,
  fn: (entry: {
    languageCode: LanguageCode;
    stringKey: string | [string, string];
    stringInLanguage: string;
  }) => void
): void {
  for (const [languageCode, uiStringsInLanguage] of Object.entries(uiStrings)) {
    assert(isLanguageCode(languageCode));
    for (const [stringKey, value] of Object.entries(uiStringsInLanguage)) {
      assert(value !== undefined);
      if (typeof value === 'string') {
        fn({
          languageCode,
          stringKey,
          stringInLanguage: value,
        });
      } else {
        for (const [stringSubKey, stringInLanguage] of Object.entries(value)) {
          assert(stringInLanguage !== undefined);
          fn({
            languageCode,
            stringKey: [stringKey, stringSubKey],
            stringInLanguage,
          });
        }
      }
    }
  }
}
