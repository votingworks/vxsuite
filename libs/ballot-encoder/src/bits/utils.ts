import { Uint8 } from './types';

/**
 * Generates a series of bitmasks in little-endian order.
 *
 * @example
 *
 * makeMasks(0) // []
 * makeMasks(1) // [0b1]
 * makeMasks(2) // [0b10, 0b01]
 * makeMasks(3) // [0b100, 0b010, 0b001]
 */
export function makeMasks<T extends number>(count: number): T[] {
  const results: T[] = [];

  for (let i = count - 1; i >= 0; i -= 1) {
    results.push((1 << i) as T);
  }

  return results;
}

/**
 * Coerces `number` to a `Uint8` by asserting it is in the range of `Uint8`.
 *
 * @throws if `number` is outside the range of `Uint8` or is not an integer
 */
export function toUint8(number: unknown): Uint8 {
  if (
    typeof number !== 'number' ||
    number < 0x00 ||
    number > 0xff ||
    (number | 0) !== number
  ) {
    throw new TypeError(`cannot convert number to Uint8: ${number}`);
  }
  return number as Uint8;
}

/**
 * Gets the size in bits of a number.
 *
 * @param number a non-negative integer
 * @returns the minimum number of bits required to represent `number`
 * @throws if `number` is negative or is not an integer
 */
export function sizeof(number: number): number {
  if (number < 0 || (number | 0) !== number) {
    throw new TypeError(
      `cannot get size of negative or non-integer: ${number}`
    );
  }

  let maxBits = 1;

  let shifted = number;
  // eslint-disable-next-line no-cond-assign
  while ((shifted >>= 1)) {
    maxBits += 1;
  }

  return maxBits;
}

/**
 * Groups `array` into arrays of size `count`.
 */
export function inGroupsOf<T>(count: number, array: T[]): Array<T[]> {
  const result: Array<T[]> = [];

  for (let i = 0; i < array.length; i += count) {
    result.push(array.slice(i, i + count));
  }

  return result;
}
