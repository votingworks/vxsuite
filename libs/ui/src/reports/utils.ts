import { find } from '@votingworks/basics';
import { Tabulation } from '@votingworks/types';

export type LabeledScannerBatch = Tabulation.ScannerBatch & { label: string };

// VxScan and VxCentralScan produce batch labels of the form 'Batch 1',
// 'Batch 2', etc., so we don't need to prefix them with 'Batch'.
export function getBatchLabel(
  batchId: string,
  scannerBatches: LabeledScannerBatch[]
): string {
  return batchId === Tabulation.MANUAL_BATCH_ID
    ? 'Manual Tallies'
    : find(scannerBatches, (batch) => batch.batchId === batchId).label;
}

export function getScannerLabel(scannerId: string): string {
  return scannerId === Tabulation.MANUAL_SCANNER_ID
    ? 'Manual Tallies'
    : scannerId;
}

export function prefixedTitle({
  isOfficial,
  isForLogicAndAccuracyTesting,
  title,
}: {
  isOfficial?: boolean;
  isForLogicAndAccuracyTesting?: boolean;
  title: string;
}): string {
  let prefix = '';
  if (isForLogicAndAccuracyTesting) {
    prefix = 'Test Deck';
  } else {
    prefix = isOfficial ? 'Official' : 'Unofficial';
  }

  return `${prefix} ${title}`;
}
