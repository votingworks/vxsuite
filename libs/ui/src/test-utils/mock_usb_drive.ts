/* istanbul ignore file */
import { throwIllegalValue } from '@votingworks/basics';

import type { UsbDriveStatus } from '@votingworks/usb-drive';

export function mockUsbDriveStatus(
  status: UsbDriveStatus['status']
): UsbDriveStatus {
  switch (status) {
    case 'mounted':
      return {
        status,
        mountPoint: 'test-mount-point',
      };
    case 'no_drive':
    case 'ejected':
      return { status };
    case 'error':
      return {
        status,
        reason: 'bad_format',
      };
    default:
      throwIllegalValue(status);
  }
}
