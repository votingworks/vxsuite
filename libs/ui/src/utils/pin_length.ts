import { assert } from '@votingworks/basics';

/**
 * Represents a range of possible PIN lengths.
 */
export class PinLength {
  private constructor(
    private readonly minimum: number,
    private readonly maximum: number
  ) {
    assert(minimum > 0, 'min must be > 0');
    assert(minimum <= maximum, 'min must be <= max');
    assert(Number.isInteger(minimum), 'min must be an integer');
    assert(Number.isInteger(maximum), 'max must be an integer');
  }

  /**
   * Create a new `PinLength` representing an inclusive range of possible PIN
   * lengths.
   *
   * @param min The minimum length of the PIN.
   * @param max The maximum length of the PIN.
   */
  static range(min: number, max: number): PinLength {
    return new PinLength(min, max);
  }

  /**
   * Create a new `PinLength` representing a fixed-length PIN.
   */
  static exactly(length: number): PinLength {
    return new PinLength(length, length);
  }

  /**
   * The minimum length of the PIN.
   */
  get min(): number {
    return this.minimum;
  }

  /**
   * The maximum length of the PIN.
   */
  get max(): number {
    return this.maximum;
  }

  /**
   * Whether the PIN length is fixed.
   */
  get isFixed(): boolean {
    return this.minimum === this.maximum;
  }
}
