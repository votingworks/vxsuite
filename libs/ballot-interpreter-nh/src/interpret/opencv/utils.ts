/* eslint-disable @typescript-eslint/no-use-before-define */

import * as cv from '@u4/opencv4nodejs';

import { err, ok, Result } from '@votingworks/types';
import { assert } from '@votingworks/utils';

import { Bit, Rect, Vector } from '../../types';
import { vec } from '../../utils';
import { GRAYSCALE_BLACK, GRAYSCALE_WHITE } from './constants';
import { MarginType, PageMargins, TimingMarks } from './types';

/**
 * Sets all pixel values below a determined brightness threshold tp white and
 * all other pixel values to black, using the Otsu threshold algorith:
 * https://en.wikipedia.org/wiki/Otsu%27s_method}
 */
export async function binarizeImage(
  image: cv.Mat,
  options: { blurSize?: number } = {}
): Promise<cv.Mat> {
  const { blurSize = 2 } = options;
  // Recommendation from openCV: make binarization a little more effective by
  // slightly blurring to remove noise:
  // FIXME: magic constants
  const denoisedImage =
    blurSize > 0
      ? await image.blurAsync(new cv.Size(blurSize, blurSize))
      : image;

  return denoisedImage.thresholdAsync(
    0,
    GRAYSCALE_WHITE,
    cv.THRESH_BINARY_INV | cv.THRESH_OTSU
  );
}

/** TODO */
export async function getShadedPixelRatio(
  image: cv.Mat,
  options: { blurSize?: number; testBounds?: cv.Rect } = {}
): Promise<number> {
  const { blurSize, testBounds } = options;

  const binarizedImage = await binarizeImage(image, { blurSize });

  const testRegion = testBounds
    ? binarizedImage.getRegion(testBounds)
    : binarizedImage;

  const numNonZeroPixels = await testRegion.countNonZeroAsync();

  return numNonZeroPixels / (testRegion.cols * testRegion.rows);
}

/** TODO */
export function decodeBitsFromBottomTimingMarks(
  timingMarks: TimingMarks
): Bit[] {
  const { detected, interpolated } = timingMarks;

  const leftBorderMark = detected[0];
  const rightBorderMark = detected[detected.length - 1];
  const bitMarks = detected.slice(1, -1).map((m) => m.boundingRect());
  const bitPlaceholders = interpolated.slice(1, -1);

  assert(leftBorderMark);
  assert(rightBorderMark);

  const decodedBits: Bit[] = Array.from<Bit>({
    length: bitPlaceholders.length,
  }).fill(0);

  let markIndex = 0;
  for (let bitIndex = 0; bitIndex < bitPlaceholders.length; bitIndex += 1) {
    if (markIndex >= bitMarks.length) {
      break;
    }

    const bitMark = bitMarks[markIndex];
    const bitPlaceholder = bitPlaceholders[bitIndex];
    assert(bitMark);
    assert(bitPlaceholder);

    if (getRectCenter(bitMark).x < bitPlaceholder.x + bitPlaceholder.width) {
      decodedBits[bitIndex] = 1;
      markIndex += 1;
    }
  }

  return decodedBits;
}

/** TODO */
export function findMinOrMaxContour(
  contours: cv.Contour[],
  query: { x: 'min' | 'max'; y: 'min' | 'max' }
): Result<cv.Contour, string> {
  if (contours.length === 0) {
    return err('contour list is empty');
  }

  let currentPick = contours[0];
  assert(currentPick);

  const xMultiplier = query.x === 'min' ? 1 : -1;
  const yMultiplier = query.y === 'min' ? 1 : -1;
  for (const candidate of contours) {
    const currentPickLocation = getRectOrigin(currentPick.boundingRect());
    const candidateLocation = getRectOrigin(candidate.boundingRect());

    const currentPickScore =
      currentPickLocation.x * xMultiplier + currentPickLocation.y * yMultiplier;
    const candidateScore =
      candidateLocation.x * xMultiplier + candidateLocation.y * yMultiplier;

    if (candidateScore < currentPickScore) {
      currentPick = candidate;
    }
  }

  return ok(currentPick);
}

/** TODO */
export function getRectCenter(r: cv.Rect): cv.Point2 {
  return new cv.Point2(r.x + r.width / 2, r.y + r.height / 2);
}

/** TODO */
export function getRectOrigin(r: cv.Rect): cv.Point2 {
  return new cv.Point2(r.x, r.y);
}

/** TODO */
export function toVxRect(r: cv.Rect): Rect {
  const centerPoint = getRectCenter(r);

  return {
    height: r.height,
    maxX: r.x + r.width,
    maxY: r.y + r.height,
    minX: r.x,
    minY: r.y,
    width: r.width,
    x: centerPoint.x,
    y: centerPoint.y,
  };
}

/** TODO */
export function toVxVector(p: cv.Point2): Vector {
  return vec(p.x, p.y);
}

/** TODO */
export function getDistancesBetweenTimingMarks(marks: cv.Rect[]): cv.Point2[] {
  const distances: cv.Point2[] = [];

  const innerMarks = marks.length > 2 ? marks.slice(1, -1) : marks;

  for (let i = 1; i < innerMarks.length; i += 1) {
    const currentMark = innerMarks[i];
    const previousMark = innerMarks[i - 1];

    assert(currentMark);
    assert(previousMark);

    distances.push(getDistanceBetweenRects(previousMark, currentMark));
  }

  return distances;
}

/** TODO */
export function translateRect(rect: cv.Rect, offset: cv.Point2): cv.Rect {
  return new cv.Rect(
    rect.x + offset.x,
    rect.y + offset.y,
    rect.width,
    rect.height
  );
}

/** Returns the distance between centers of the given rectangles. */
export function getDistanceBetweenRects(from: cv.Rect, to: cv.Rect): cv.Point2 {
  return getRectCenter(to).sub(getRectCenter(from));
}

/** TODO */
export async function mapMargins<U, T>(
  margins: PageMargins<U>,
  fn: (margin: U, marginType: MarginType) => Promise<T>
): Promise<PageMargins<T>> {
  const [bottom, left, right, top] = await Promise.all([
    fn(margins.bottom, 'bottom'),
    fn(margins.left, 'left'),
    fn(margins.right, 'right'),
    fn(margins.top, 'top'),
  ]);

  return {
    bottom,
    left,
    right,
    top,
  };
}

/** TODO */
export async function padImage(
  image: cv.Mat,
  padding: Partial<PageMargins<number>>,
  color = GRAYSCALE_BLACK
): Promise<cv.Mat> {
  return await image.copyMakeBorderAsync(
    padding.top ?? 0,
    padding.bottom ?? 0,
    padding.left ?? 0,
    padding.right ?? 0,
    cv.BORDER_CONSTANT,
    color
  );
}

/** TODO */
export function average(...values: number[]): number {
  return (
    values.reduce((totalSoFar, value) => totalSoFar + value) / values.length
  );
}

/** TODO */
export function clamp(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}

/**
 * Estimates the area within the convex bounds of the given contour that are
 * filled. Not as accurate as counting the non-zero pixels in a binarized image
 * region, but useful for distinguishing text from a timing mark, for example.
 */
export function estimateContourFillRatio(contour: cv.Contour): number {
  return Math.abs(contour.area) / Math.abs(contour.convexHull().area);
}
