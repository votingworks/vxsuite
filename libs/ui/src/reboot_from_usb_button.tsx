import { LogEventId, Logger } from '@votingworks/logging';
import { assert, throwIllegalValue } from '@votingworks/utils';
import React, { useCallback, useEffect, useState } from 'react';
import styled from 'styled-components';
import { Button } from './button';
import { UsbDriveStatus } from './hooks/use_usb_drive';
import { Loading } from './loading';
import { Modal } from './modal';
import { Prose } from './prose';

enum State {
  CLOSED = 'closed',
  NO_USB_INSERTED = 'no_usb_inserted',
  NO_BOOTABLE_USB_OPTION = 'no_bootable_usb_option',
  PREPARING_BOOT = 'preparing_boot',
  REBOOTING = 'rebooting',
}

const UsbImage = styled.img`
  margin-right: auto;
  margin-left: auto;
  height: 200px;
`;

interface Props {
  usbDriveStatus: UsbDriveStatus;
  logger: Logger;
}

/**
 * Button that attempts to reboot the machine from a mounted, bootable USB drive if one exists.
 * Shows a prompt to insert a USB to the user when loaded with no mounted USB.
 * Prompts the user to reboot and try again when loaded with a mounted USB but window.kiosk.prepareToBootFromUsb returns false indicating there is no bootable usb option.
 */

export function RebootFromUsbButton({
  usbDriveStatus,
  logger,
}: Props): JSX.Element {
  const [currentState, setCurrentState] = useState(State.CLOSED);
  const attemptReboot = useCallback(async () => {
    assert(window.kiosk);
    await logger.log(
      LogEventId.PrepareBootFromUsbInit,
      'system_administrator',
      {
        message:
          'Attempting to set the USB drive next in the boot order and reboot',
      }
    );
    if (usbDriveStatus !== 'mounted') {
      setCurrentState(State.NO_USB_INSERTED);
      await logger.log(
        LogEventId.PrepareBootFromUsbComplete,
        'system_administrator',
        {
          message: 'No USB drive found, could not boot from usb',
          disposition: 'failure',
        }
      );
      return;
    }
    setCurrentState(State.PREPARING_BOOT);
    const readyToReboot = await window.kiosk.prepareToBootFromUsb();
    if (readyToReboot) {
      await logger.log(
        LogEventId.PrepareBootFromUsbComplete,
        'system_administrator',
        {
          message:
            'USB Drive successfully set next in the boot order for the next boot.',
          disposition: 'success',
        }
      );
      await logger.log(LogEventId.RebootMachine, 'system_administrator', {
        message: 'Machine will now reboot…',
      });
      setCurrentState(State.REBOOTING);
      await window.kiosk.reboot();
    } else {
      setCurrentState(State.NO_BOOTABLE_USB_OPTION);
      await logger.log(
        LogEventId.PrepareBootFromUsbComplete,
        'system_administrator',
        {
          message:
            'USB Drive not found in the list of bootable options, boot order is not modified.',
          disposition: 'failure',
        }
      );
    }
  }, [usbDriveStatus, logger]);
  async function reboot() {
    assert(window.kiosk);
    await logger.log(LogEventId.RebootMachine, 'system_administrator', {
      message: 'User trigged a reboot of the machine…',
    });
    setCurrentState(State.REBOOTING);
    await window.kiosk.reboot();
  }
  useEffect(() => {
    if (
      usbDriveStatus === 'mounted' &&
      currentState === State.NO_USB_INSERTED
    ) {
      void attemptReboot();
    }
  }, [usbDriveStatus, currentState, attemptReboot]);

  const onClose = useCallback(
    () => setCurrentState(State.CLOSED),
    [setCurrentState]
  );

  if (currentState === State.PREPARING_BOOT) {
    return <Modal content={<Loading>Preparing to boot…</Loading>} />;
  }
  if (currentState === State.REBOOTING) {
    return <Modal content={<Loading>Rebooting…</Loading>} />;
  }
  if (currentState === State.NO_BOOTABLE_USB_OPTION) {
    return (
      <Modal
        content={
          <Prose>
            The USB Drive was not found in the list of bootable devices. Make
            sure it is the correct USB drive and reboot the machine with it
            plugged in, then try again.
          </Prose>
        }
        onOverlayClick={onClose}
        actions={
          <React.Fragment>
            <Button primary onPress={reboot}>
              Reboot
            </Button>
            <Button onPress={onClose}>Close</Button>
          </React.Fragment>
        }
      />
    );
  }
  if (currentState === State.NO_USB_INSERTED) {
    return (
      <Modal
        content={
          <Prose>
            <h1>No USB Drive Detected</h1>
            <p>
              <UsbImage src="/assets/usb-drive.svg" alt="Insert USB Image" />
              Please insert a USB drive to boot from.
            </p>
          </Prose>
        }
        onOverlayClick={onClose}
        actions={<Button onPress={onClose}>Cancel</Button>}
      />
    );
  }
  if (currentState === State.CLOSED) {
    return <Button onPress={attemptReboot}>Reboot from USB</Button>;
  }
  /* istanbul ignore next - compile time check for completeness */
  throwIllegalValue(currentState);
}
