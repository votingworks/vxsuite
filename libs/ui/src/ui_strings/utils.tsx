import React from 'react';

import {
  BallotStyleId,
  Candidate,
  Election,
  Parties,
  Precinct,
  PrecinctSelection,
  getCandidateParties,
  getPartyForBallotStyle,
} from '@votingworks/types';
import { getPrecinctSelection } from '@votingworks/utils';

import { appStrings } from './app_strings';
import { electionStrings } from './election_strings';

/**
 * Convenience component for rendering a translated list of parties associated
 * with a given candidate, along with the relevant audio.
 */
export function CandidatePartyList(props: {
  candidate: Candidate;
  electionParties: Parties;
}): JSX.Element {
  const { candidate, electionParties } = props;

  return (
    <React.Fragment>
      {getCandidateParties(electionParties, candidate).map((party, i) => (
        <React.Fragment key={party.id}>
          {/*
           * TODO(kofi): This comma-delimiting isn't properly
           * internationalized (comma character is rendered differently in
           * different languages/character sets) -- need to figure out a clean
           * way to do this.
           */}
          {i > 0 && <React.Fragment>, </React.Fragment>}
          {electionStrings.partyName(party)}
        </React.Fragment>
      ))}
    </React.Fragment>
  );
}

export function PrecinctSelectionName(props: {
  electionPrecincts: readonly Precinct[];
  precinctSelection?: PrecinctSelection;
}): React.ReactNode {
  const { electionPrecincts, precinctSelection } = props;

  if (!precinctSelection) {
    return null;
  }

  if (precinctSelection.kind === 'AllPrecincts') {
    return appStrings.labelAllPrecinctsSelection();
  }

  const precinct = getPrecinctSelection(electionPrecincts, precinctSelection);

  return electionStrings.precinctName(precinct);
}

export function PrimaryElectionTitlePrefix(props: {
  ballotStyleId: BallotStyleId;
  election: Election;
}): React.ReactNode {
  const { ballotStyleId, election } = props;
  const party = getPartyForBallotStyle({ ballotStyleId, election });
  if (!party) {
    return null;
  }

  return (
    <React.Fragment>{electionStrings.partyFullName(party)} </React.Fragment>
  );
}
