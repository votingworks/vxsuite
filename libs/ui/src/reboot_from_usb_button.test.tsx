import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import { usbstick } from '@votingworks/utils';
import { fakeKiosk } from '@votingworks/test-utils';
import { Logger, LogSource } from '@votingworks/logging';
import { RebootFromUsbButton } from './reboot_from_usb_button';

beforeEach(() => {
  window.kiosk = fakeKiosk();
});

test('renders without a USB drive as expected.', async () => {
  render(
    <RebootFromUsbButton
      usbDriveStatus={usbstick.UsbDriveStatus.absent}
      logger={new Logger(LogSource.VxAdminFrontend)}
    />
  );
  // Initially should just contain the button
  fireEvent.click(screen.getByText('Reboot from USB'));
  await screen.findByText('No USB Drive Detected');
});

test('renders with a non-bootable USB as expected', async () => {
  window.kiosk!.prepareToBootFromUsb = jest.fn().mockResolvedValue(false);
  render(
    <RebootFromUsbButton
      usbDriveStatus={usbstick.UsbDriveStatus.mounted}
      logger={new Logger(LogSource.VxAdminFrontend)}
    />
  );
  fireEvent.click(screen.getByText('Reboot from USB'));
  await screen.findByText(
    /The USB Drive was not found in the list of bootable devices./
  );
  expect(window.kiosk!.prepareToBootFromUsb).toHaveBeenCalledTimes(1);
  expect(window.kiosk!.reboot).toHaveBeenCalledTimes(0);
  fireEvent.click(screen.getByText('Close'));
  expect(
    screen.queryAllByText(
      /The USB Drive was not found in the list of bootable devices./
    )
  ).toHaveLength(0);
  fireEvent.click(screen.getByText('Reboot from USB'));
  await screen.findByText(
    /The USB Drive was not found in the list of bootable devices./
  );
  fireEvent.click(screen.getByText('Reboot'));
  await screen.findByText('Rebooting…');
  expect(window.kiosk!.reboot).toHaveBeenCalledTimes(1);
});

test('reboots automatically when clicked with a bootable USB', async () => {
  window.kiosk!.prepareToBootFromUsb = jest.fn().mockResolvedValue(true);
  render(
    <RebootFromUsbButton
      usbDriveStatus={usbstick.UsbDriveStatus.mounted}
      logger={new Logger(LogSource.VxAdminFrontend)}
    />
  );
  fireEvent.click(screen.getByText('Reboot from USB'));
  await screen.findByText('Rebooting…');
  expect(window.kiosk!.prepareToBootFromUsb).toHaveBeenCalledTimes(1);
  expect(window.kiosk!.reboot).toHaveBeenCalledTimes(1);
});

test('modal state updates when USB drive is inserted.', async () => {
  window.kiosk!.prepareToBootFromUsb = jest.fn().mockResolvedValue(false);
  const { rerender } = render(
    <RebootFromUsbButton
      usbDriveStatus={usbstick.UsbDriveStatus.absent}
      logger={new Logger(LogSource.VxAdminFrontend)}
    />
  );
  fireEvent.click(screen.getByText('Reboot from USB'));
  await screen.findByText('No USB Drive Detected');
  rerender(
    <RebootFromUsbButton
      usbDriveStatus={usbstick.UsbDriveStatus.mounted}
      logger={new Logger(LogSource.VxAdminFrontend)}
    />
  );
  await screen.findByText(/The USB Drive was not found/);
  expect(window.kiosk!.prepareToBootFromUsb).toHaveBeenCalledTimes(1);
  expect(window.kiosk!.reboot).toHaveBeenCalledTimes(0);
});
