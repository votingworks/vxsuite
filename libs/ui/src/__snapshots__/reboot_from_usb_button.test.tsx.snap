// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without a USB drive as expected. 1`] = `
.c0 {
  display: inline-block;
  border: none;
  border-radius: 0.25em;
  box-shadow: 0 0 0 0 rgba(71,167,75,1);
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 0.75em 1em;
  text-align: center;
  line-height: 1.25;
  color: black;
  touch-action: manipulation;
}

.c0:hover,
.c0:active {
  outline: none;
}

<div>
  <button
    class="c0"
    type="button"
  >
    Reboot from USB
  </button>
</div>
`;
